src/TextUI/TestRunner.php

```diff
@@ -174,14 +176,31 @@ public function doRun(Test $suite, array $arguments = [], bool $exit = true): Te
             \mt_srand($arguments['randomOrderSeed']);
         }
 
-        if ($arguments['executionOrder'] !== TestSuiteSorter::ORDER_DEFAULT || $arguments['resolveDependencies']) {
-            $sorter = new TestSuiteSorter;
+        if (isset($arguments['cacheResultFile'])) {
+            $cache = new TestResultCache($arguments['cacheResultFile']);
+        }
+
+        if ($arguments['executionOrder'] !== TestSuiteSorter::ORDER_DEFAULT || $arguments['executionOrderDefects'] !== TestSuiteSorter::ORDER_DEFAULT || $arguments['resolveDependencies']) {
+            if (!isset($cache)) {
+                $cache = new TestResultCache;
+            }
+            $cache->load();
+            $sorter = new TestSuiteSorter($cache);
 
-            $sorter->reorderTestsInSuite($suite, $arguments['executionOrder'], $arguments['resolveDependencies']);
+            $sorter->reorderTestsInSuite($suite, $arguments['executionOrder'], $arguments['resolveDependencies'], $arguments['executionOrderDefects']);
 
             unset($sorter);
         }
 
+        if ($arguments['cacheResult']) {
+            if (isset($arguments['cacheResultFile'])) {
+                $outputCache = new TestResultCache($arguments['cacheResultFile']);
```

Comment:
do we really need to distinguish `$cache` from `$outputCache` ?
can't we use same cache instance ?

Also, if one would pass `cacheResultFile` but no `cacheResult`, he will get (input) `$cache`, but no `$outputCache`.
I see this problematic and I would not allow for that

Reply:
Agreed. The edge case in the PHPUnit self-tests has been solved differently. I can now make this cleaner, too.